src/state_manager/mod.rs

```diff
@@ -1230,53 +1252,77 @@ where
-            last_state = st;
-            last_receipt = msg_root;
-        }
-        Ok(())
+    /// Validates all tipsets at epoch `start..=end` behind the heaviest tipset.
+    ///
+    /// This spawns [`rayon::current_num_threads`] threads to do the compute-heavy work
+    /// of tipset validation.
+    ///
+    /// # What is validation?
+    /// Every state transition returns a new _state root_, which is typically retained in e.g snapshots.
+    /// For "full" snapshots, all state roots are retained.
+    /// For standard snapshots, the last 2000 or so state roots are retained.
+    ///
+    /// _receipts_ meanwhile, are typically ephemeral, but each tipset knows the _receipt root_
+    /// (hash) of the previous tipset.
+    ///
+    /// This function takes advantage of that fact to validate tipsets:
+    /// - `tipset[N]` claims that `receipt_root[N-1]` should be `0xDEADBEEF`
+    /// - find `tipset[N-1]`, and perform its state transition to get the actual `receipt_root`
+    /// - assert that they match
+    ///
+    /// See [`Self::compute_tipset_state_blocking`] for an explanation of state transitions.
+    ///
+    /// # Known issues
+    /// This function is blocking, but we do observe threads waiting and synchronizing.
+    /// This is suspected to be due to the VM's use of the chain store, or maybe JIT-ing.
+    #[tracing::instrument(skip(self))]
+    pub fn validate(self: &Arc<Self>, epochs: RangeInclusive<i64>) -> anyhow::Result<()> {
+        use rayon::iter::ParallelIterator as _;
+
+        let heaviest = self.cs.heaviest_tipset();
+        let heaviest_epoch = heaviest.epoch();
+        let end = self
+            .cs
+            .tipset_by_height(*epochs.end(), heaviest, false)
+            .context(format!(
+            "couldn't get a tipset at height {} behind heaviest tipset at height {heaviest_epoch}",
+            *epochs.end(),
+        ))?;
+
+        // lookup tipset parents as we go along, iterating DOWN from `end`
+        let tipsets = itertools::unfold(Some(end), |tipset| {
+            let child = tipset.take()?;
+            // if this has parents, unfold them in the next iteration
+            *tipset = self.cs.tipset_from_keys(child.parents()).ok();
+            Some(child)
+        });
+
+        tipsets
+            .take_while(|tipset| tipset.epoch() >= *epochs.start())
+            .tuple_windows()
+            .par_bridge()
+            .try_for_each(|(child, parent)| {
+                info!(height = parent.epoch(), "compute parent state");
+                let (actual_state, actual_receipt) = self
+                    .compute_tipset_state_blocking(parent, NO_CALLBACK)
+                    .context("couldn't compute tipset state")?;
+                let expected_receipt = child.min_ticket_block().message_receipts();
+                let expected_state = child.parent_state();
+                match (expected_state, expected_receipt) == (&actual_state, &actual_receipt) {
```

Comment:
Once again, why not a simple `if-else`? What is the advantage of this verbosity and an increased indentation level?

Reply:
Same as here:

https://github.com/ChainSafe/forest/pull/2893#discussion_r1220358284